### ISSUE TITLE
y_inline won't work after few gmx [linux]

### ISSUE THREAD

**Issue by kadaradam** [pawn-lang/YSI-Includes#99]
Tittle says everything. My server restarts every 5 hour, and after few server restarts (gmx), the y_inline stops working, and non of the dialogs give response. 

I tried to reproduce it on windows, so i restarted the server like 12 times and the dialogs always worked. So I'm not sure where the problem is.

**Comment by Y-Less** [comment]
Yes, people have said this before, and every time the report boils down to "this doesn't work, I don't know why."  When someone actually gives me code that reliably demonstrates the problem, I will fix it, until then - I can't reproduce it and have nothing more to go on!

In fact, I think there are still a few similar open issues on this.  If I had powers on this repository I'd have closed them all as either "more info required" or "cannot reproduce".  Fortunately for you, @Misiur is nicer than I am.

**Comment by kadaradam** [comment]
I can't provide more info, maybe if you send me the include with debug messages, I can compile that with the gamemode, and then send you the results.

There are no warning or error messages, the y_inline dialogs just does not work after few server restarts. I'm running Southclaw's [ScavengeSurvive](https://github.com/Southclaw/ScavengeSurvive) gamemode.

**Comment by Y-Less** [comment]
What include?

**Comment by kadaradam** [comment]
y_inline with debugging

**Comment by Y-Less** [comment]
I'm not sure what you want sending though.

**Comment by kadaradam** [comment]
I think I was clear, but never mind.
I enabled debugging in the [y_debug.inc](https://github.com/Misiur/YSI-Includes/blob/YSI.tl/YSI_Core/y_debug.inc#L127), will that information help to fix the issue if I send you the debug messages?

**Comment by Y-Less** [comment]
Yes, that's all the debugging there is - there is no special y_inline include with debugging, just the normal one.

**Comment by Y-Less** [comment]
However, I think you are the first person to explicitly mention that they are using Linux, and which GM they are using - which may be another step in the right direction for me tracking down the problem.  Someone else gave some useful information in another issue, so I'll try and combine them to see what I come up with.  As I say, step number one is reproduction, since I've never seen the bug myself it makes it very hard to fix.

**Comment by Southclaws** [comment]
I will have to look into this too, SS issue is here: Southclaw/ScavengeSurvive#248 I haven't actually noticed this either but I don't currently run my own server and do all my dev on windows. I do a _lot_ of logging in when testing certain things since I send a gmx after every build, then use the login dialog so it sounds like it's a linux specific thing for sure.

**Comment by kadaradam** [comment]
I know and understand it, but it's very hard to reproduce this issue, since sometimes works, simetimes not. Yesterday I had this issue 2 times, today it worked so far. That's why I asked you previously to add more debug message to the y_inline include, so I can send you back the results when it fails to work.

Any idea why I don't get any debug message? I also added **`state ysi_debug : on;`** to `OnGameModeInit()`

```
    DebugLevel(7);
    logf("[LOGPROMPT] %p is logging in", playerid);

    inline Response(pid, dialogid, response, listitem, string:inputtext[])
    {
        #pragma unused pid, dialogid, listitem

        if(response)
        {
            if(strlen(inputtext) < 4)
            {
                acc_LoginAttempts[playerid]++;

                if(acc_LoginAttempts[playerid] < 5)
                {
                    DisplayLoginPrompt(playerid, 1);
                }
                else
                {
                    ChatMsgAll(GREY, " >  %p left the server without logging in.", playerid);
                    Kick(playerid);
                }

                return 1;
            }

            new
                inputhash[MAX_PASSWORD_LEN],
                storedhash[MAX_PASSWORD_LEN];

            WP_Hash(inputhash, MAX_PASSWORD_LEN, inputtext);
            GetPlayerPassHash(playerid, storedhash);

            if(!strcmp(inputhash, storedhash))
            {
                Login(playerid);
            }
            else
            {
                acc_LoginAttempts[playerid]++;

                if(acc_LoginAttempts[playerid] < 5)
                {
                    DisplayLoginPrompt(playerid, 1);
                }
                else
                {
                    ChatMsgAll(GREY, " >  %p left the server without logging in.", playerid);
                    Kick(playerid);
                }
            }
        }
        else
        {
            ChatMsgAll(GREY, " >  %p left the server without logging in.", playerid);
            Kick(playerid);
        }

        return 1;
    }
    Dialog_ShowCallback(playerid, using inline Response, DIALOG_STYLE_PASSWORD, "Login To Your Account", str, "Accept", "Leave");

    DebugLevel(0);
```

EDIT:
Oh, i just realized that I have to set the level to -1 with `_DEBUG` to use `DebugLevel()`. Let me check that.

EDIT 2:
That worked, now I'll upload the script to the host, and wait till the dialogs stop working.

**Comment by HxTheDash** [comment]
Not only on Linux that this happens.
I use Windows 7 ultimate 64 bits and encounter the same issue.

**Comment by kadaradam** [comment]
Weird, but since I opened this issue and added debugs to the gamemode, I haven't encounter the issue.

**Comment by Y-Less** [comment]
test

**Comment by kadaradam** [comment]
It happend again, but I just found a strange thing.

```
[12/06/2016 21:17:03] [LOAD] <UNKNOWN> (account does not exist)
[12/06/2016 21:17:03] [REGPROMPT] <UNKNOWN> is registering
```

UNKNOWN is a non existing account, also no player were online at that time.
Maybe the issue is server related, but it's weird that everything works, except dialogs.

Anyway here are the logs, when the dialogs worked fine:

```
First dialog after server restart:
[09/06/2016 21:01:22] [LOAD] Preston_Marlowe (account exists, prompting login) Alive: 1 Last login: Thursday Jun 09 2016 at 20:50:32
[09/06/2016 21:01:22] [LOGPROMPT] Preston_Marlowe is logging in
[09/06/2016 21:01:22] Inline_Entry called: Response:....
[09/06/2016 21:01:22] AMX_Read called: 0 -100
[09/06/2016 21:01:22] AMX_RawRead called: 0
[09/06/2016 21:01:22] AMX_Write called: 201898204, 2
[09/06/2016 21:01:22] AMX_Write called: 201898208, 948300
[09/06/2016 21:01:22] AMX_Write called: 201898212, 32770
[09/06/2016 21:01:22] AMX_Write called: 201898220, 0
[09/06/2016 21:01:22] Inline_Allocator called
[09/06/2016 21:01:22] Inline_Main called (6)
[09/06/2016 21:01:22] Inline_Main: YSI_g_sCurInlineEntry = E784C
[09/06/2016 21:01:22] Inline_Main: YSI_g_sCurInlineLoop = E78CC
[09/06/2016 21:01:22] Inline_Main: YSI_g_sCurInlineCode = E7990
[09/06/2016 21:01:22] Inline_Main: varCount = 4
[09/06/2016 21:01:22] Inline_Main: arrCount = 0
[09/06/2016 21:01:22] Inline_Main: strCount = 1
[09/06/2016 21:01:22] Inline_Main: TYPE 2
[09/06/2016 21:01:22] AMX_Write called: -2097976, -1511317140
[09/06/2016 21:01:22] AMX_Write called: 201898216, 948444
[09/06/2016 21:01:22] AMX_Write called: 201898204, 2
[09/06/2016 21:01:22] Callback_Get called: ResponseLÜ
[09/06/2016 21:01:22] Malloc_Allocate called: 133, 60509676, 60509675, 12105912
[09/06/2016 21:01:22] Callback_Get end
[09/06/2016 21:01:26] [LOGIN] Preston_Marlowe logged in

An another login dialog:
[09/06/2016 23:13:56] [connection] 1:63177 requests connection cookie.
[09/06/2016 23:13:57] [connection] incoming connection: 1:63177 id: 0
[09/06/2016 23:13:58] [join] JohnW has joined the server (0:1)
[09/06/2016 23:13:58] INFO: [RemoveObjects_FirstLoad] Created session data for JohnW
[09/06/2016 23:13:58] [JOIN] JohnW joined
[09/06/2016 23:14:03] [LOAD] JohnW (account exists, prompting login) Alive: 1 Last login: Tuesday Jun 07 2016 at 20:38:10
[09/06/2016 23:14:03] [LOGPROMPT] JohnW is logging in
[09/06/2016 23:14:03] Callback_Get called: ResponseLÜ
[09/06/2016 23:14:03] Malloc_Allocate called: 133, 60516262, 60509675, 12099326
[09/06/2016 23:14:03] Callback_Get end
[09/06/2016 23:14:06] [LOGIN] JohnW logged in
```

And when the dialogs didn't work:

```
First dialog after server restart:
[12/06/2016 21:22:47] [connection] 1:15491 requests connection cookie.
[12/06/2016 21:22:48] [connection] incoming connection: 1:15491 id: 0
[12/06/2016 21:22:48] [join] YOLO has joined the server (0:1)
[12/06/2016 21:22:48] INFO: [RemoveObjects_FirstLoad] Created session data for YOLO
[12/06/2016 21:22:48] [JOIN] YOLO joined
[12/06/2016 21:22:53] [LOAD] YOLO (account exists, prompting login) Alive: 1 Last login: Sunday Jun 12 2016 at 21:11:31
[12/06/2016 21:22:53] [LOGPROMPT] YOLO is logging in
[12/06/2016 21:22:53] Inline_Entry called: Response:....
[12/06/2016 21:22:53] AMX_Write called: 201898548, 2
[12/06/2016 21:22:53] AMX_Write called: 201898552, 950188
[12/06/2016 21:22:53] AMX_Write called: 201898556, 32770
[12/06/2016 21:22:53] AMX_Write called: 201898564, 201897336
[12/06/2016 21:22:53] Inline_Allocator called
[12/06/2016 21:22:53] Inline_Main called (6)
[12/06/2016 21:22:53] Inline_Main: YSI_g_sCurInlineEntry = E7FAC
[12/06/2016 21:22:53] Inline_Main: YSI_g_sCurInlineLoop = E802C
[12/06/2016 21:22:53] Inline_Main: YSI_g_sCurInlineCode = E80F0
[12/06/2016 21:22:53] Inline_Main: varCount = 4
[12/06/2016 21:22:53] Inline_Main: arrCount = 0
[12/06/2016 21:22:53] Inline_Main: strCount = 1
[12/06/2016 21:22:53] Inline_Main: TYPE 2
[12/06/2016 21:22:53] AMX_Write called: -2101012, -1855043380
[12/06/2016 21:22:53] AMX_Write called: 201898560, 950332
[12/06/2016 21:22:53] AMX_Write called: 201898548, 2
[12/06/2016 21:22:53] Callback_Get called: Response¬<x
[12/06/2016 21:22:53] Malloc_Allocate called: 133, 60509814, 60509813, 0
[12/06/2016 21:23:21] INFO: [OnPlayerDisconnect] Removing session data file for YOLO
[12/06/2016 21:23:21] [LOGOUT] YOLO not logged in.
[12/06/2016 21:23:21] [part] YOLO has left the server (0:1)

Second dialog:
[12/06/2016 21:51:49] [connection] 1:56367 requests connection cookie.
[12/06/2016 21:51:50] [connection] incoming connection: 1:56367 id: 0
[12/06/2016 21:51:51] [join] Preston_Marlowe has joined the server (0:1)
[12/06/2016 21:51:51] INFO: [RemoveObjects_FirstLoad] Created session data for Preston_Marlowe
[12/06/2016 21:51:51] [JOIN] Preston_Marlowe joined
[12/06/2016 21:51:56] [LOAD] Preston_Marlowe (account exists, prompting login) Alive: 1 Last login: Sunday Jun 12 2016 at 16:05:51
[12/06/2016 21:51:56] [LOGPROMPT] Preston_Marlowe is logging in
[12/06/2016 21:51:56] Callback_Get called: Response¬<x
[12/06/2016 21:51:56] Malloc_Allocate called: 133, 60509814, 60509813, 0
[12/06/2016 21:52:26] [chat] [Preston_Marlowe]: dialog is not working
[12/06/2016 21:53:44] INFO: [OnPlayerDisconnect] Removing session data file for Preston_Marlowe
[12/06/2016 21:53:44] [LOGOUT] Preston_Marlowe not logged in.
[12/06/2016 21:53:44] [part] Preston_Marlowe has left the server (0:1)
```

If you don't find any ordinary in the debug messages, I'll try to look inside the gamemode more deeply.

**Comment by Southclaws** [comment]
Sorry, the `<UNKNOWN>` thing happens when a player disconnects during the `LoadAccountDelay` period, that deferred function should be aborted on disconnect and contain an IsPlayerConnected check. I don't think this could be the cause though since I'm pretty sure everything is caught and handled gracefully despite the player ID being invalid.

**Comment by kadaradam** [comment]
Yeah, I managed to reproduce it after i posted my previous comment, and dialogs worked fine. So that should not cause the dialogs not to work.

**Comment by Y-Less** [comment]
Hopefully fixed by #101 when using `YSI_NO_HEAP_MALLOC` as a compile-time setting.  This reverts y_malloc to the old style of using a huge array instead of the heap, but will increase compile times and file sizes.
